### PR TITLE
Refactor YOLOX

### PR DESCRIPTION
**Pull request by hhaAndroid**
## Motivation

The official code has been updated recently. In order to align accuracy, we have also reconstructed

- Alignment mAP
- Change ema param 0.0002 to 0.0001
- Add YOLOXHSVRandomAug
- Optimize unnecessary `all_reduce_dict` code
- Skip unnecessary filtering rules
- reduce mean the num_total_samples

## BC-breaking (Optional)

None

### CODE REVIEW

File: mmdet/datasets/dataset_wrappers.py [hunk 1]
```diff
@@ -311,6 +312,8 @@ def __init__(self,
                  pipeline,
                  dynamic_scale=None,
                  skip_type_keys=None):
+        warnings.warn('dynamic_scale is deprecated. will be removed '
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:315]: If raising warning here, the _dynamic_scale function should not be removed.

Comment (RangiLyu) [reply mmdet/datasets/dataset_wrappers.py:315]: If this function is no longer supported, please raise an error and give some information, like "Please use another method..." or something.

File: mmdet/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -1983,20 +1983,16 @@ class Mosaic:
...
            output. Default to (0.5, 1.5).
-        min_bbox_size (int | float): The minimum pixel for filtering
-            invalid bboxes after the mosaic pipeline. Default to 0.
         pad_val (int): Pad value. Default to 114.
     """
 
     def __init__(self,
                  img_scale=(640, 640),
                  center_ratio_range=(0.5, 1.5),
-                 min_bbox_size=0,
```
Comment (RangiLyu) [inline mmdet/datasets/pipelines/transforms.py:1994]: need to add warning and tell users to use another method

File: mmdet/datasets/pipelines/transforms.py [hunk 3]
```diff
@@ -2513,15 +2435,14 @@ def __call__(self, results):
...
 
         # Translation
-        trans_x = random.uniform(0.5 - self.max_translate_ratio,
-                                 0.5 + self.max_translate_ratio) * width
-        trans_y = random.uniform(0.5 - self.max_translate_ratio,
-                                 0.5 + self.max_translate_ratio) * height
+        trans_x = random.uniform(self.max_translate_ratio,
+                                 self.max_translate_ratio) * width
+        trans_y = random.uniform(self.max_translate_ratio,
+                                 self.max_translate_ratio) * height
```
Comment (RangiLyu) [inline mmdet/datasets/pipelines/transforms.py:2441]: Is this correct?

File: mmdet/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -2468,10 +2401,7 @@ def __init__(self,
                  scaling_ratio_range=(0.5, 1.5),
                  max_shear_degree=2.0,
                  border=(0, 0),
-                 border_val=(114, 114, 114),
-                 min_bbox_size=2,
-                 min_area_ratio=0.2,
-                 max_aspect_ratio=20):
```
Comment (RangiLyu) [inline mmdet/datasets/pipelines/transforms.py:2474]: BC-Breaking

File: mmdet/core/evaluation/eval_hooks.py [hunk 5]
```diff
@@ -22,7 +59,48 @@ def _do_evaluate(self, runner):
             self._save_ckpt(runner, key_score)
 
 
-class DistEvalHook(BaseDistEvalHook):
+class DistEvalHook(EvalHook):
```
Comment (RangiLyu) [inline mmdet/core/evaluation/eval_hooks.py:62]: Can this hook support distributed training?

File: mmdet/core/evaluation/eval_hooks.py [hunk 6]
```diff
@@ -1,13 +1,50 @@
...
+                self.dynamic_values.append(dynamic_interval[1])
+
+    def before_train_epoch(self, runner):
+        """Evaluate the model only at the start of training by epoch."""
+        if self.dynamic_intervals and self.by_epoch:
+            step = bisect.bisect(self.dynamic_steps, (runner.epoch + 1))
+            # Dynamically modify the evaluation interval
+            self.interval = self.dynamic_values[step-1]
+
+        if not (self.by_epoch and self.initial_flag):
```
Comment (ZwwWayne) [inline mmdet/core/evaluation/eval_hooks.py:31]: for line in 31-35, we can use super().before_train_epoch(runner)

File: mmdet/core/evaluation/eval_hooks.py [hunk 7]
```diff
@@ -1,13 +1,50 @@
...
+            self.dynamic_steps = [0]
+            self.dynamic_values = [self.interval]
+            for dynamic_interval in dynamic_intervals:
+                assert isinstance(dynamic_interval, tuple)
+                self.dynamic_steps.append(dynamic_interval[0])
+                self.dynamic_values.append(dynamic_interval[1])
+
+    def before_train_epoch(self, runner):
+        """Evaluate the model only at the start of training by epoch."""
+        if self.dynamic_intervals and self.by_epoch:
```
Comment (ZwwWayne) [inline mmdet/core/evaluation/eval_hooks.py:26]: extract line 26-29 to a function like self.decide_interval

File: mmdet/models/detectors/yolox.py [hunk 8]
```diff
@@ -15,6 +22,81 @@ def __init__(self,
                  train_cfg=None,
                  test_cfg=None,
                  pretrained=None,
-                 init_cfg=None):
+                 init_cfg=None,
```
Comment (RangiLyu) [inline mmdet/models/detectors/yolox.py:25]: Suggest to put init_cfg at last.

File: mmdet/models/detectors/yolox.py [hunk 9]
```diff
@@ -15,6 +22,81 @@ def __init__(self,
...
+                 input_size=(640, 640),
+                 random_size_range=(15, 25),
+                 random_size_interval=10):
```
Comment (RangiLyu) [inline mmdet/models/detectors/yolox.py:28]: Docstring

File: mmdet/models/detectors/yolox.py [hunk 10]
```diff
@@ -15,6 +22,81 @@ def __init__(self,
...
+        # Multi-scale training
+        img, gt_bboxes = self.preprocess(img, gt_bboxes)
+
+        losses = super(YOLOX, self).forward_train(img, img_metas, gt_bboxes,
+                                                  gt_labels, gt_bboxes_ignore)
+
+        # random resizing
+        if (self.progress_in_iter + 1) % self.random_size_interval == 0:
+            self.input_size = self.random_resize()
+        self.progress_in_iter += 1
```
Comment (RangiLyu) [inline mmdet/models/detectors/yolox.py:70]: Using an internal counter or using a hook? Which is a better design?

File: mmdet/core/evaluation/eval_hooks.py [hunk 11]
```diff
@@ -1,13 +1,50 @@
...
 from torch.nn.modules.batchnorm import _BatchNorm
 
 
 class EvalHook(BaseEvalHook):
+    def __init__(self, *args, dynamic_intervals=None, **kwargs):
+        super(EvalHook, self).__init__(*args, **kwargs)
+
+        self.dynamic_intervals = dynamic_intervals
+        if dynamic_intervals is not None:
+            assert isinstance(dynamic_intervals, list)
```
Comment (ZwwWayne) [inline mmdet/core/evaluation/eval_hooks.py:16]: assert is_list_of(dynamic_intervals, tuple)

**Review by RangiLyu — COMMENTED** [review]
Need to fix random affine unit tests.

File: mmdet/core/evaluation/eval_hooks.py [hunk 12]
```diff
@@ -1,13 +1,50 @@
...
+            self.interval = self.dynamic_values[step-1]
+
+        if not (self.by_epoch and self.initial_flag):
+            return
+        if self.start is not None and runner.epoch >= self.start:
+            self.after_train_epoch(runner)
+        self.initial_flag = False
+
+    def before_train_iter(self, runner):
+        if self.dynamic_intervals and not self.by_epoch:
```
Comment (ZwwWayne) [inline mmdet/core/evaluation/eval_hooks.py:38]: use self.decide_interval

File: mmdet/core/evaluation/eval_hooks.py (hunk 5 above)
Comment (hhaAndroid) [reply mmdet/core/evaluation/eval_hooks.py:62]: Yes

File: mmdet/models/detectors/yolox.py (hunk 10 above)
Comment (hhaAndroid) [reply mmdet/models/detectors/yolox.py:70]: The hook cannot get the img object. My opinion is that for the time being, I will switch to resize later.

File: configs/yolox/yolox_s_8x8_300e_coco.py [hunk 13]
```diff
@@ -116,27 +129,27 @@
...
         interval=interval,
         priority=48),
     dict(type='ExpMomentumEMAHook', resume_from=resume_from, priority=49)
 ]
 checkpoint_config = dict(interval=interval)
-evaluation = dict(interval=interval, metric='bbox')
+evaluation = dict(
+    save_best='auto',
+    interval=interval,
+    dynamic_intervals=[(max_epochs - num_last_epochs, 1)],
```
Comment (RangiLyu) [inline configs/yolox/yolox_s_8x8_300e_coco.py:153]: Setting these args at the same time is a little bit confusing here. Can we only use `dynamic_intervals` here and add some comments?

File: mmdet/core/evaluation/eval_hooks.py [hunk 14]
```diff
@@ -1,14 +1,47 @@
...
+            assert mmcv.is_list_of(dynamic_intervals, tuple)
+
+            self.dynamic_steps = [0]
+            self.dynamic_steps.extend([
+                dynamic_interval[0] for dynamic_interval in dynamic_intervals
+            ])
+            self.dynamic_values = [self.interval]
+            self.dynamic_values.extend([
+                dynamic_interval[1] for dynamic_interval in dynamic_intervals
+            ])
```
Comment (RangiLyu) [inline mmdet/core/evaluation/eval_hooks.py:27]: The name `dynamic_steps` and `dynamic_values` are not clear enough.
```suggestion
        self.use_dynamic_intervals = dynamic_intervals is not None
        if self.use_dynamic_intervals:
            assert mmcv.is_list_of(dynamic_intervals, tuple)

            self.dynamic_milestones = [0]
            self.dynamic_milestones.extend([
                dynamic_interval[0] for dynamic_interval in dynamic_intervals
            ])
            self.dynamic_intervals = [self.interval]
            self.dynamic_intervals.extend([
                dynamic_interval[1] for dynamic_interval in dynamic_intervals
            ])
```

File: mmdet/core/hook/ema.py [hunk 15]
```diff
@@ -17,7 +17,7 @@ class BaseEMAHook(Hook):
         momentum (float): The momentum used for updating ema parameter.
             Ema's parameter are updated with the formula:
            `ema_param = (1-momentum) * ema_param + momentum * cur_param`.
-            Defaults to 0.0002.
+            Defaults to 0.0001.
```
Comment (RangiLyu) [inline mmdet/core/hook/ema.py:20]: Changing the default value may cause BC-breaking. Suggest changing this value in config.

File: mmdet/core/hook/yolox_mode_switch_hook.py [hunk 16]
```diff
@@ -34,5 +34,9 @@ def before_train_epoch(self, runner):
         if (epoch + 1) == runner.max_epochs - self.num_last_epochs:
             runner.logger.info('No mosaic and mixup aug now!')
             train_loader.dataset.update_skip_type_keys(self.skip_type_keys)
+            if hasattr(train_loader, 'persistent_workers'
```
Comment (RangiLyu) [inline mmdet/core/hook/yolox_mode_switch_hook.py:37]: Need to add comments.

File: mmdet/datasets/pipelines/loading.py [hunk 17]
```diff
@@ -550,18 +550,24 @@ class FilterAnnotations:
             boxes.
     """
 
-    def __init__(self, min_gt_bbox_wh):
+    def __init__(self, min_gt_bbox_wh, always_keep=False):
```
Comment (RangiLyu) [inline mmdet/datasets/pipelines/loading.py:553]: What does `always_keep` mean?

File: configs/yolox/yolox_s_8x8_300e_coco.py (hunk 13 above)
Comment (hhaAndroid) [reply configs/yolox/yolox_s_8x8_300e_coco.py:153]: There will be bc breaking.

Comment (hhaAndroid) [reply configs/yolox/yolox_s_8x8_300e_coco.py:153]: This will cause conflict

File: mmdet/datasets/pipelines/loading.py (hunk 17 above)
Comment (hhaAndroid) [reply mmdet/datasets/pipelines/loading.py:553]: I didn't expect a better name. Its meaning is that when the filtering rule becomes an empty box, the filtering rule is skipped at this time, and the result before the filter is always retained.

File: mmdet/models/detectors/yolox.py [hunk 18]
```diff
@@ -15,6 +28,81 @@ def __init__(self,
                  train_cfg=None,
                  test_cfg=None,
                  pretrained=None,
+                 input_size=(640, 640),
+                 random_size_range=(15, 25),
```
[the remark this answers is not in the document]
Comment (hhaAndroid) [reply mmdet/models/detectors/yolox.py:32]: Its logic is to first randomize and then multiply by 32, so the real image size cannot be used directly. But you can add a coefficient.

File: configs/yolox/README.md [hunk 19]
```diff
@@ -17,9 +17,13 @@
 
 | Backbone  | size   | Mem (GB) |   box AP | Config | Download |
 |:---------:|:-------:|:-------:|:-------:|:--------:|:------:|
+| YOLOX-s | 640 |   7.9      |   40.2  | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolox/yolox_s_8x8_300e_coco.py)       |[model](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_s_8x8_300e_coco/yolox_s_8x8_300e_coco_20211110_150543-699b132f.pth) &#124; [log](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_s_8x8_300e_coco/yolox_s_8x8_300e_coco_20211110_150543.log.json) |
```
Comment (RangiLyu) [inline configs/yolox/README.md:20]: The score threshold of the mAP is 0.001?

**Review by RangiLyu — APPROVED** [review]

File: configs/yolox/README.md (hunk 19 above)
Comment (hhaAndroid) [reply configs/yolox/README.md:20]: Yes

**Comment by yulinhuyang** [comment]
hoping this request will be merged quickly ,I want to have a test in yoloX_s!

File: configs/yolox/README.md [hunk 20]
```diff
@@ -17,9 +17,13 @@
...
+| YOLOX-l | 640 |   19.6      |   49.4  | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolox/yolox_l_8x8_300e_coco.py)       |[model](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_l_8x8_300e_coco/yolox_l_8x8_300e_coco_20211113_121356-5e33d212.pth) &#124; [log](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_l_8x8_300e_coco/yolox_l_8x8_300e_coco_20211113_121356.log.json) |
 | YOLOX-Tiny | 416 |   3.6      |   31.6  | [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolox/yolox_tiny_8x8_300e_coco.py)       |[model](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_tiny_8x8_300e_coco/yolox_tiny_8x8_300e_coco_20210806_234250-4ff3b67e.pth) &#124; [log](https://download.openmmlab.com/mmdetection/v2.0/yolox/yolox_tiny_8x8_300e_coco/yolox_tiny_8x8_300e_coco_20210806_234250.log.json) |
 
+
 **Note**:
 
 1. The test score threshold is 0.001.
-2. We find that the performance is unstable and may fluctuate by about 0.7 mAP. We will continue to investigate and improve it.
+2. YOLOX-s and YOLOX-l are the models corresponding to the best map on the validation set.
```
Comment (ZwwWayne) [inline configs/yolox/README.md:28]: are from the checkpoints having the best mAP on the validation set rather than the last checkpoints in experiments.

File: configs/yolox/README.md [hunk 21]
```diff
@@ -17,9 +17,13 @@
...
+3. We find that the performance is unstable and may fluctuate by about 0.7 mAP. We will continue to investigate and improve it.
```
Comment (ZwwWayne) [inline configs/yolox/README.md:29]: continue investigating it.

File: configs/yolox/yolox_s_8x8_300e_coco.py [hunk 22]
```diff
@@ -74,16 +71,19 @@
         transforms=[
             dict(type='Resize', keep_ratio=True),
             dict(type='RandomFlip'),
-            dict(type='Pad', size=img_scale, pad_val=114.0),
-            dict(type='Normalize', **img_norm_cfg),
+            dict(
+                type='Pad',
+                pad_to_square=True,
+                pad_val=dict(img=(114.0, 114.0, 114.0))),
```
Comment (ZwwWayne) [inline configs/yolox/yolox_s_8x8_300e_coco.py:77]: Add comments to indicate the necessity

File: mmdet/core/evaluation/eval_hooks.py [hunk 23]
```diff
@@ -1,14 +1,52 @@
...
 import os.path as osp
 
+import mmcv
 import torch.distributed as dist
 from mmcv.runner import DistEvalHook as BaseDistEvalHook
 from mmcv.runner import EvalHook as BaseEvalHook
 from torch.nn.modules.batchnorm import _BatchNorm
 
 
+def _calc_dynamic_intervals(start_interval, dynamic_interval_list):
```
Comment (ZwwWayne) [inline mmdet/core/evaluation/eval_hooks.py:12]: make it an attribute of EvalHook

File: mmdet/datasets/pipelines/loading.py (hunk 17 above)
Comment (ZwwWayne) [reply mmdet/datasets/pipelines/loading.py:553]: how about keep_empty?

File: mmdet/datasets/pipelines/transforms.py [hunk 24]
```diff
@@ -1985,18 +1985,23 @@ class Mosaic:
            output. Default to (0.5, 1.5).
         min_bbox_size (int | float): The minimum pixel for filtering
             invalid bboxes after the mosaic pipeline. Default to 0.
+        skip_filter_rule (bool): Whether to skip filtering rules. If it
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:1988]: skip_filter might be better

File: mmdet/models/detectors/yolox.py [hunk 25]
```diff
@@ -1,12 +1,47 @@
...
 
 
 @DETECTORS.register_module()
 class YOLOX(SingleStageDetector):
     r"""Implementation of `YOLOX: Exceeding YOLO Series in 2021
-    <https://arxiv.org/abs/2107.08430>`_"""
+    <https://arxiv.org/abs/2107.08430>`_
+
+    Note: Multi-scale training in the model is not recommended.
+    Taking into account the training speed and accuracy, it will
```
Comment (ZwwWayne) [inline mmdet/models/detectors/yolox.py:19]: Considering the trade-off between training speed and accuracy, multi-scale training is temporarily kept.
More elegant implementation will be adopted in the future.

File: mmdet/datasets/pipelines/loading.py (hunk 17 above)
Comment (hhaAndroid) [reply mmdet/datasets/pipelines/loading.py:553]: OK

**Review by ZwwWayne — APPROVED** [review]
